File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        // Get the first node
        ListNode* head = nullptr;
        if(list1 != nullptr && list2 != nullptr) {
            if(list1->val < list2->val) {
                head = list1;
                list1 = list1->next;
            }
            else {
                head = list2;
                list2 = list2->next;
            }
        }
        else if(list1 != nullptr) {
            return list1;
        }
        else {
            return list2;
        }
        // Store the merged list
        ListNode* merged_list = head;
        while(list1 != nullptr && list2 != nullptr) {
            if(list1->val < list2->val) {
                merged_list->next = list1;
                merged_list = list1;
                list1 = list1->next;
            }
            else {
                merged_list->next = list2;
                merged_list = list2;
                list2 = list2->next;
            }
        }
        if(list1 != nullptr) {
            merged_list->next = list1;
        }
        else if(list2 != nullptr) {
            merged_list->next = list2;
        }
        return head;
    }
};

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    """Solution for detecting cycles in a singly-linked list."""

    def hasCycle(self, head: Optional[ListNode]) -> bool:
        """Check if a linked list contains a cycle.

        Uses Floyd's cycle-finding algorithm (tortoise and hare) with two
        pointers moving at different speeds. If a cycle exists, the fast
        pointer will eventually meet the slow pointer.

        Args:
            head: The head node of the linked list.

        Returns:
            True if the linked list has a cycle, False otherwise.
        """
        if head is None:
            return False
        
        slow: Optional[ListNode] = head
        fast: Optional[ListNode] = head

        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                return True

        return False
